sql: allowing using the same PK CONSTRAINT NAME when paving over rowid

otan opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/73906

Previously, `ADD CONSTRAINT PRIMARY KEY` with the matching postgres
primary key name `tbl_pkey` worked as we had a different PRIMARY KEY
name and thus no conflicts with the existing implicit rowid PK.

However, when we changed PRIMARY KEYs to have the correct name,
this functionality was broken.

This commit fixes that by allowing the same expected PostgreSQL PRIMARY
KEY name to pave over the implicit rowid PRIMARY KEY.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73833)
<!-- Reviewable:end -->

otan committed:
sql: allowing using the same PK CONSTRAINT NAME when paving over rowid

Previously, `ADD CONSTRAINT PRIMARY KEY` with the matching postgres
primary key name `tbl_pkey` worked as we had a different PRIMARY KEY
name and thus no conflicts with the existing implicit rowid PK.

However, when we changed PRIMARY KEYs to have the correct name,
this functionality was broken.

This commit fixes that by allowing the same expected PostgreSQL PRIMARY
KEY name to pave over the implicit rowid PRIMARY KEY.

Release note: None

rafiss commented:
what would the 21.2 behavior be if you tried doing `ALTER TABLE t ADD CONSTRAINT "primary" PRIMARY KEY (id)`?

otan commented:
you'd conflict and error

otan commented:
oh wait, it's fine in v21.1 :0

```

demo@127.0.0.1:26257/defaultdb> create table t (id int not null);
CREATE TABLE

Time: 4ms total (execution 4ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> ALTER TABLE t ADD CONSTRAINT "primary" PRIMARY KEY (id);
NOTICE: primary key changes are finalized asynchronously; further schema changes on this table may be restricted until the job completes
ALTER TABLE

Time: 82ms total (execution 82ms / network 0ms)
```

rafiss commented:
Nice find, yeah seems possible that https://github.com/cockroachdb/cockroach/commit/e89093faf7225a635ddb5d4f260b20a4d8dd64e2 changed it unintentionally.

Also looks like there was a test for this behavior in 4e504dc74a, but maybe that test got broken by 53a58186f79def90f66a9fb461600683c34c5a0c

rafiss commented:
So I think we should go ahead with this PR, but we should also figure out if this code is not relevant: https://github.com/cockroachdb/cockroach/blob/dd7fd4e5a371ce4437fce383b9674cba2e212fe6/pkg/sql/alter_table.go#L1626-L1636

otan commented:
> but we should also figure out if this code is not relevant

the code tests that you can't add a `PRIMARY KEY` constraint when you already add one, e.g.

```
demo@127.0.0.1:26257/defaultdb>
CREATE TABLE t (id int primary key, b int not null);
ALTER TABLE t ADD CONSTRAINT bpkey PRIMARY KEY (b);
CREATE TABLE

Time: 7ms total (execution 6ms / network 0ms)

ERROR: multiple primary keys for table "t" are not allowed
SQLSTATE: 42P16
```

rafiss reviewed (approved):
lgtm! please make sure sql-schema signs off for awareness

ajwerner commented on pkg/sql/logictest/testdata/logic_test/alter_table:
```diff
@@ -2014,3 +2014,34 @@ ALTER TABLE t ALTER COLUMN b SET DEFAULT 10
+)
+
+statement error constraint with name "t_pkey" already exists
+ALTER TABLE t ADD CONSTRAINT t_pkey PRIMARY KEY (id)
```
Our abstractions around `IsPrimaryIndexDefaultRowID` are pretty crappy. I feel like the decision to always show these automatically created elements but then to detect whether they are the automatically created elements in an ad-hoc way is sort of annoying. Like what allows us to trigger this special behavior of deleting `t_pk` and adding the new primary column is just that we have a hidden column that's a the primary key column and there's only one. Like what the heck is that about? Imagine I did this:

```sql
CREATE TABLE t (
  id INT NOT NULL,
  my_beloved_column INT8 NOT VISIBLE NOT NULL,
  CONSTRAINT t_pk_that_provides_nice_uniqueness PRIMARY KEY (my_beloved_column),
)
```
```
> alter table t add primary key (id);
NOTICE: primary key changes are finalized asynchronously; further schema changes on this table may be restricted until the job completes
ALTER TABLE


Time: 136ms total (execution 136ms / network 0ms)

demo@127.0.0.1:63578/defaultdb> show create table t;
  table_name |                 create_statement
-------------+---------------------------------------------------
  t          | CREATE TABLE public.t (
             |     id INT8 NOT NULL,
             |     my_beloved_column INT8 NOT VISIBLE NOT NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (id ASC),
             |     FAMILY "primary" (id, my_beloved_column)
             | )
```

My feeling is that this code is bad mostly because `IsPrimaryIndexDefaultRowID` is bad.

otan replied:
Yeah I mean unless we store a bool for detecting this or add a check that rowid is a prefix of the col name not sure we can do better. Anything you want me to action here?

ajwerner replied:
We could look at the type, the name, and the name of the primary key constraint. Thoughts?

otan replied:
My only hang up with PK constraint name is renaming the table. Rest I think is agreeable

ajwerner replied:
> My only hang up with PK constraint name is renaming the table. Rest I think is agreeable

That's legit. Also the fact that we renamed what we call it now. How about also the existence of a default value for the thing. It'd be weird for it to not have a default value.

otan replied:
all done

otan committed:
tabledesc: tighten definition check on IsPrimaryIndexDefaultRowID

Release note: None

otan commented:
bors r=rafiss
